File: clients/R/dataset_client.R
# Sample R helper for ColonyCore Dataset Service API
# Requires the httr and jsonlite packages.

library(httr)
library(jsonlite)

cc_dataset_headers <- function(token = NULL) {
  headers <- add_headers(
    `User-Agent` = "colonycore-dataset-client/0.1",
    Accept = "application/json"
  )
  if (!is.null(token) && nzchar(token)) {
    headers <- add_headers(headers, Authorization = paste("Bearer", token))
  }
  headers
}

cc_list_templates <- function(base_url, token = NULL, timeout = 30) {
  url <- paste0(rtrim(base_url), "/api/v1/datasets/templates")
  resp <- GET(url, cc_dataset_headers(token), timeout(timeout))
  stop_for_status(resp)
  content(resp, as = "parsed", simplifyVector = TRUE)$templates
}

cc_get_template <- function(
  base_url,
  plugin,
  key,
  version,
  token = NULL,
  timeout = 30
) {
  url <- sprintf(
    "%s/api/v1/datasets/templates/%s/%s/%s",
    rtrim(base_url),
    plugin,
    key,
    version
  )
  resp <- GET(url, cc_dataset_headers(token), timeout(timeout))
  stop_for_status(resp)
  content(resp, as = "parsed", simplifyVector = TRUE)$template
}

cc_validate_template <- function(
  base_url,
  plugin,
  key,
  version,
  parameters = list(),
  token = NULL,
  timeout = 30
) {
  url <- sprintf(
    "%s/api/v1/datasets/templates/%s/%s/%s/validate",
    rtrim(base_url),
    plugin,
    key,
    version
  )
  resp <- POST(
    url,
    cc_dataset_headers(token),
    timeout(timeout),
    body = list(parameters = parameters),
    encode = "json"
  )
  stop_for_status(resp)
  content(resp, as = "parsed", simplifyVector = TRUE)
}

cc_run_template <- function(
  base_url,
  plugin,
  key,
  version,
  parameters = list(),
  scope = list(),
  format = "json",
  token = NULL,
  timeout = 60
) {
  url <- sprintf(
    "%s/api/v1/datasets/templates/%s/%s/%s/run",
    rtrim(base_url),
    plugin,
    key,
    version
  )
  query <- list(format = tolower(format))
  headers <- cc_dataset_headers(token)
  if (tolower(format) == "csv") {
    headers <- add_headers(headers, Accept = "text/csv")
  }
  resp <- POST(
    url,
    headers,
    timeout(timeout),
    query = query,
    body = list(parameters = parameters, scope = scope),
    encode = "json"
  )
  stop_for_status(resp)
  if (tolower(format) == "csv") {
    return(content(resp, as = "text", encoding = "UTF-8"))
  }
  content(resp, as = "parsed", simplifyVector = TRUE)
}

cc_queue_export <- function(
  base_url,
  template_slug = NULL,
  plugin = NULL,
  key = NULL,
  version = NULL,
  parameters = list(),
  scope = list(),
  formats = character(),
  requested_by = NULL,
  reason = NULL,
  project_id = NULL,
  protocol_id = NULL,
  token = NULL,
  timeout = 30
) {
  template <- if (!is.null(template_slug) && nzchar(template_slug)) {
    list(slug = template_slug)
  } else {
    if (any(!nzchar(c(plugin, key, version)))) {
      stop(
        "plugin, key, and version must be supplied when slug is omitted",
        call. = FALSE
      )
    }
    list(plugin = plugin, key = key, version = version)
  }
  body <- list(
    template = template,
    parameters = parameters,
    scope = scope,
    formats = as.list(formats),
    requested_by = requested_by,
    reason = reason,
    project_id = project_id,
    protocol_id = protocol_id
  )
  url <- paste0(rtrim(base_url), "/api/v1/datasets/exports")
  resp <- POST(
    url,
    cc_dataset_headers(token),
    timeout(timeout),
    body = body,
    encode = "json"
  )
  stop_for_status(resp)
  content(resp, as = "parsed", simplifyVector = TRUE)$export
}

cc_get_export <- function(base_url, export_id, token = NULL, timeout = 30) {
  url <- sprintf("%s/api/v1/datasets/exports/%s", rtrim(base_url), export_id)
  resp <- GET(url, cc_dataset_headers(token), timeout(timeout))
  stop_for_status(resp)
  content(resp, as = "parsed", simplifyVector = TRUE)$export
}

cc_wait_for_export <- function(
  base_url,
  export_id,
  token = NULL,
  poll_seconds = 2,
  timeout = 300
) {
  deadline <- Sys.time() + timeout
  repeat {
    export <- cc_get_export(base_url, export_id, token = token)
    if (export$status %in% c("succeeded", "failed")) {
      return(export)
    }
    if (Sys.time() > deadline) {
      stop(sprintf("export %s timed out", export_id), call. = FALSE)
    }
    Sys.sleep(poll_seconds)
  }
}

cc_download_artifact <- function(url, path = NULL, token = NULL, timeout = 60) {
  resp <- GET(url, cc_dataset_headers(token), timeout(timeout))
  stop_for_status(resp)
  if (is.null(path)) {
    return(content(resp, as = "raw"))
  }
  dir.create(dirname(path), showWarnings = FALSE, recursive = TRUE)
  writeBin(content(resp, as = "raw"), path)
  invisible(path)
}

rtrim <- function(x) {
  sub("/*$", "", x)
}

File: clients/python/dataset_client.py
"""Sample Python helper for the ColonyCore Dataset Service.

This helper maps to the OpenAPI contract defined in docs/schema/dataset-service.openapi.yaml
and demonstrates how analysts can script dataset access and exports from their own runtime.
"""

from __future__ import annotations

import json
import os
import time
import warnings
from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List, Optional

import requests

_DEFAULT_HEADERS = {
    "Accept": "application/json",
    "User-Agent": "colonycore-dataset-client/0.1",
}
_MAX_TEMPLATE_PAGES = 1000
_DATASET_SCOPE_HEADERS = {
    "requestor": "X-Dataset-Requestor",
    "roles": "X-Dataset-Roles",
    "project_ids": "X-Dataset-Project-Ids",
    "protocol_ids": "X-Dataset-Protocol-Ids",
}
_STREAM_PROGRESS_HEADER = "X-Progress"


@dataclass
class ExportHandle:
    """Represents an asynchronous export job."""

    id: str
    status: str
    progress_pct: int = 0
    eta_seconds: Optional[int] = None
    progress_state: Optional[str] = None
    artifact_readiness: Optional[str] = None
    artifacts: List[Dict[str, Any]] = field(default_factory=list)
    error: Optional[str] = None
    template: Dict[str, Any] = field(default_factory=dict)
    scope: Dict[str, Any] = field(default_factory=dict)
    parameters: Dict[str, Any] = field(default_factory=dict)
    formats: List[str] = field(default_factory=list)
    requested_by: Optional[str] = None
    reason: Optional[str] = None
    project_id: Optional[str] = None
    protocol_id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    completed_at: Optional[str] = None


@dataclass
class CSVRunResult:
    """Represents a streamed CSV response plus transport metadata."""

    body: str
    progress: Optional[str]
    headers: Dict[str, str] = field(default_factory=dict)


class DatasetClient:
    """Thin convenience wrapper around the DatasetService REST API."""

    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        timeout: int = 30,
        session: Optional[requests.Session] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout
        self._session = session or requests.Session()
        for header, value in _DEFAULT_HEADERS.items():
            self._session.headers.setdefault(header, value)
        if api_key:
            self._session.headers.setdefault("Authorization", f"Bearer {api_key}")

    # Template helpers -----------------------------------------------------
    def _scope_headers(self, scope: Optional[Dict[str, Any]]) -> Dict[str, str]:
        if not scope:
            return {}

        headers: Dict[str, str] = {}
        for key, header in _DATASET_SCOPE_HEADERS.items():
            value = scope.get(key)
            if value is None or value == "":
                continue
            if isinstance(value, (list, tuple, set)):
                items = [str(item).strip() for item in value if str(item).strip()]
                if items:
                    headers[header] = ",".join(items)
                continue
            headers[header] = str(value)
        return headers

    def _export_handle(self, payload: Dict[str, Any]) -> ExportHandle:
        return ExportHandle(
            id=payload["id"],
            status=payload["status"],
            progress_pct=payload.get("progress_pct", 0),
            eta_seconds=payload.get("eta_seconds"),
            progress_state=payload.get("progress_state"),
            artifact_readiness=payload.get("artifact_readiness"),
            artifacts=list(payload.get("artifacts") or []),
            error=payload.get("error"),
            template=dict(payload.get("template") or {}),
            scope=dict(payload.get("scope") or {}),
            parameters=dict(payload.get("parameters") or {}),
            formats=list(payload.get("formats") or []),
            requested_by=payload.get("requested_by"),
            reason=payload.get("reason"),
            project_id=payload.get("project_id"),
            protocol_id=payload.get("protocol_id"),
            created_at=payload.get("created_at"),
            updated_at=payload.get("updated_at"),
            completed_at=payload.get("completed_at"),
        )

    def list_templates_page(
        self,
        page: int = 1,
        page_size: int = 50,
        scope: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        resp = self._session.get(
            f"{self._base_url}/api/v1/datasets/templates",
            params={"page": page, "page_size": page_size},
            headers=self._scope_headers(scope),
            timeout=self._timeout,
        )
        resp.raise_for_status()
        return resp.json()

    def list_templates(
        self,
        page: int = 1,
        page_size: int = 50,
        scope: Optional[Dict[str, Any]] = None,
    ) -> List[Dict[str, Any]]:
        templates: List[Dict[str, Any]] = []
        current_page = page
        pages_fetched = 0

        while True:
            payload = self.list_templates_page(page=current_page, page_size=page_size, scope=scope)
            pages_fetched += 1
            page_templates = payload.get("templates", [])
            if not page_templates:
                break

            templates.extend(page_templates)
            if not (payload.get("pagination") or {}).get("has_next"):
                break
            if pages_fetched >= _MAX_TEMPLATE_PAGES:
                warnings.warn(
                    f"template listing truncated after {_MAX_TEMPLATE_PAGES} pages; additional pages were not fetched",
                    RuntimeWarning,
                    stacklevel=2,
                )
                break

            current_page += 1

        return templates

    def get_template(self, plugin: str, key: str, version: str) -> Dict[str, Any]:
        path = f"{self._base_url}/api/v1/datasets/templates/{plugin}/{key}/{version}"
        resp = self._session.get(path, timeout=self._timeout)
        resp.raise_for_status()
        return resp.json()["template"]

    def validate_template(
        self,
        plugin: str,
        key: str,
        version: str,
        parameters: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        path = f"{self._base_url}/api/v1/datasets/templates/{plugin}/{key}/{version}/validate"
        resp = self._session.post(path, json={"parameters": parameters or {}}, timeout=self._timeout)
        resp.raise_for_status()
        return resp.json()

    def run_template(
        self,
        plugin: str,
        key: str,
        version: str,
        parameters: Optional[Dict[str, Any]] = None,
        scope: Optional[Dict[str, Any]] = None,
        output_format: str = "json",
        include_stream_metadata: bool = False,
    ) -> Any:
        path = f"{self._base_url}/api/v1/datasets/templates/{plugin}/{key}/{version}/run"
        headers = {}
        fmt = output_format.lower()
        if fmt == "csv":
            headers["Accept"] = "text/csv"
        payload = {
            "parameters": parameters or {},
            "scope": scope or {},
        }
        resp = self._session.post(path, params={"format": fmt}, json=payload, headers=headers, timeout=self._timeout)
        resp.raise_for_status()
        if fmt == "csv":
            if include_stream_metadata:
                return CSVRunResult(
                    body=resp.text,
                    progress=resp.headers.get(_STREAM_PROGRESS_HEADER),
                    headers=dict(resp.headers),
                )
            return resp.text
        return resp.json()

    # Export helpers -------------------------------------------------------
    def queue_export(
        self,
        template_slug: Optional[str] = None,
        *,
        plugin: Optional[str] = None,
        key: Optional[str] = None,
        version: Optional[str] = None,
        parameters: Optional[Dict[str, Any]] = None,
        scope: Optional[Dict[str, Any]] = None,
        formats: Optional[Iterable[str]] = None,
        requested_by: Optional[str] = None,
        reason: Optional[str] = None,
        project_id: Optional[str] = None,
        protocol_id: Optional[str] = None,
    ) -> ExportHandle:
        template: Dict[str, Any]
        if template_slug:
            template = {"slug": template_slug}
        else:
            if not all([plugin, key, version]):
                raise ValueError("plugin, key, and version are required when slug is omitted")
            template = {"plugin": plugin, "key": key, "version": version}
        body = {
            "template": template,
            "parameters": parameters or {},
            "scope": scope or {},
            "formats": list(formats or []),
            "requested_by": requested_by,
            "reason": reason,
            "project_id": project_id,
            "protocol_id": protocol_id,
        }
        resp = self._session.post(f"{self._base_url}/api/v1/datasets/exports", json=body, timeout=self._timeout)
        resp.raise_for_status()
        payload = resp.json()["export"]
        return self._export_handle(payload)

    def get_export(self, export_id: str) -> ExportHandle:
        resp = self._session.get(f"{self._base_url}/api/v1/datasets/exports/{export_id}", timeout=self._timeout)
        resp.raise_for_status()
        payload = resp.json()["export"]
        return self._export_handle(payload)

    def wait_for_export(
        self,
        export_id: str,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
    ) -> ExportHandle:
        deadline = time.time() + timeout
        handle = self.get_export(export_id)
        while handle.status not in {"succeeded", "failed"}:
            if time.time() > deadline:
                raise TimeoutError(f"export {export_id} did not complete within {timeout} seconds")
            time.sleep(poll_interval)
            handle = self.get_export(export_id)
        return handle

    def download_artifact(self, url: str, destination: Optional[str] = None) -> str:
        """Download an artifact using the signed URL returned by the export service."""
        resp = self._session.get(url, timeout=self._timeout)
        resp.raise_for_status()
        if destination is None:
            return resp.text
        os.makedirs(os.path.dirname(destination) or ".", exist_ok=True)
        mode = "wb" if isinstance(resp.content, (bytes, bytearray)) else "w"
        with open(destination, mode) as fh:
            fh.write(resp.content if "b" in mode else resp.text)
        return destination


def dump_pretty(obj: Any) -> str:
    """Render response payloads as formatted JSON for notebooks or logs."""
    return json.dumps(obj, indent=2, sort_keys=True)
